vectorial_model.py

```diff
@@ -0,0 +1,141 @@
+
+
+def get_normalized_term_frequency(term, doc_id, index_frequency, stats_collection):
+    """Get normalized term frequency"""
+    term_frequency = get_term_frequency(term, doc_id, index_frequency)
+    normalized_term_frequency = 0.5 + 0.5 * (term_frequency / stats_collection[doc_id]["max_frequency"])
+    return normalized_term_frequency
+
+
+def get_normalized_logarithmic_term_frequency(term, doc_id, index_frequency, stats_collection):
+    """Get normalized logarithmic term frequency"""
+    term_frequency = get_term_frequency(term, doc_id, index_frequency)
+    normalized_logarithmic_term_frequency = (1 + math.log(term_frequency)) / (1 + math.log(
+        stats_collection[doc_id]["average_frequency"]))
+    return normalized_logarithmic_term_frequency
+
+
+def get_idf(term, index_frequency, nb_doc):
+    """Gets idf of term in collection"""
+    return math.log(nb_doc / len(index_frequency[term].keys()))
+
+
+def get_stats_document(document):
+    """
+    Get statistics on the document.
+
+    Calculates the frequency of most common term, the number of unique terms, and the average frequency of terms.
+    :param document: Document to get statistics on
+    :return: (dict) stats
+    """
+    counter = Counter()
+    for term in document:
+        counter.update([term])
+    stats = dict()
+    stats["max_frequency"] = counter.most_common(1)[0][1]
+    stats["number_unique_terms"] = len(counter.items())
+    tf_moy = sum(counter.values())
+    stats["average_frequency"] = tf_moy / len(counter.items())
+    return stats
+
+
+def get_stats_collection(collection):
+    """Get statistics on all the documents in the collection """
+    stats = dict()
+    stats["nb_docs"] = len(collection.keys())
+    for doc in collection:
+        stats[doc] = get_stats_document(collection[doc])
+    return stats
+
+
+def process_vectorial_query(query, frequency_index, stop_words, stats_collection, weighting_scheme_document,
+                            weighting_scheme_query):
+    # Set weighting schemes if not specified
+    if weighting_scheme_document is None:
+        weighting_scheme_document = "tf_idf_normalize"
+    if weighting_scheme_query is None:
+        weighting_scheme_query = "frequency"
+
+    relevant_docs = {}
+    counter_query = Counter()
+    # Preprocess query
+    query_pre_processed = pre_processed_query(query, frequency_index, stop_words)
+
+    nb_doc = stats_collection["nb_docs"]
+    norm_query = 0.
+    for term in query_pre_processed:
+        if term in frequency_index:
+            w_term_query = 0.
+            counter_query.update([term])
+            if weighting_scheme_query == "binary":
+                w_term_query = 1
+            if weighting_scheme_query == "frequency":
+                w_term_query = counter_query[term]
+            norm_query = norm_query + w_term_query * w_term_query
+            for doc in frequency_index[term]:
+                w_term_doc = 0.
+                relevant_docs[doc] = 0.
```

Comment:
I am not sure but I think this could be an error.
If you look below (line 139), it seems we want to increment `relevant_docs[doc]`:
```
relevant_docs[doc] = relevant_docs[doc] + w_term_doc * w_term_query
```
So why are we setting it to 0 every time we find the doc ?

Reply:
This is also taken from the lab, I think line 125 can just be deleted.

Reply:
Deleted line 125 and used this `relevant_docs[doc] = relevant_docs.get(doc, 0) + w_term_doc * w_term_query`